## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when R is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [58]:
type("my_thing")
len("my_thing")

8

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [59]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", '"', "(", ")", " ", "-", ";", ":"]

2. Make one long list containing all the symbols.  Find the length of this list.


In [60]:
all_symbols = lower_case + upper_case + punctuation

len(all_symbols)

64

In [61]:
all_symbols

['a',
 'b',
 'c',
 'd',
 'e',
 'f',
 'g',
 'h',
 'i',
 'j',
 'k',
 'l',
 'm',
 'n',
 'o',
 'p',
 'q',
 'r',
 's',
 't',
 'u',
 'v',
 'w',
 'x',
 'y',
 'z',
 'A',
 'B',
 'C',
 'D',
 'E',
 'F',
 'G',
 'H',
 'I',
 'J',
 'K',
 'L',
 'M',
 'N',
 'O',
 'P',
 'Q',
 'R',
 'S',
 'T',
 'U',
 'V',
 'W',
 'X',
 'Y',
 'Z',
 '.',
 ',',
 '!',
 '?',
 "'",
 '"',
 '(',
 ')',
 ' ',
 '-',
 ';',
 ':']

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [62]:
import numpy as np

lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate([lower_case, upper_case, punctuation])

all_symbols, len(all_symbols)

(array(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
        'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
        'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
        '.', ',', '!', '?', "'", '"', '(', ')', ' ', '-', ';', ':'],
       dtype='<U1'),
 64)

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [63]:
import pandas as pd

symbols_df = pd.DataFrame(all_symbols)

symbols_df = symbols_df.rename(columns = {0: "Symbol"})

display(symbols_df)

Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]

In [64]:
# import pandas as pd

# symbols_df = pd.DataFrame(all_symbols, columns=["Symbol"])

## Part Two: Decoding the secret message.


In [120]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret = top_secret[0].to_numpy()


In [121]:
top_secret = np.asarray(top_secret)

In [122]:
t_s = ((top_secret + 14) * 18) - 257

t_s

array([ 2.03876872,  1.9560115 ,  1.68364791,  1.09861229,  1.68364791,
        2.14522972,  1.38629436,  1.76318026,  1.64791843,  1.73286795,
        1.76318026,  1.74825378,  1.74825378,  2.20335962,  2.25542975,
        1.74825378,  1.93560051,  1.84443973,  2.29755993,  2.18472393,
        1.9560115 ,  1.94591015,  2.33641442,  2.10234631,  2.06356719,
        2.05543693,  2.26629975,  2.07944154,  2.2094203 ,  2.39789527,
        2.30756026,  2.18472393,  2.23866841,  2.14522972,  2.24431818,
        2.44517456,  2.45263739,  4.12098328,  3.44275484,  2.54070218,
        2.29248374,  4.12203564,  2.45263739,  3.28474071,  3.01293299,
        2.38108697,  4.12334797,  2.28735549,  2.84178988,  2.3767951 ,
        4.12439537,  2.51521896,  2.61555431,  2.81200875,  2.36809922,
        2.77258872,  2.46723697,  4.12622305,  2.64665241,  2.87787111,
        4.1270043 ,  3.70274783,  2.62351204,  2.51521896,  4.12804407,
        3.24687692,  2.45263739,  2.47082121,  2.5468751 ,  4.17

In [123]:
import numpy as np
t_s = np.exp(t_s)

t_s

array([ 7.68114575,  7.07106781,  5.38516481,  3.        ,  5.38516481,
        8.54400375,  4.        ,  5.83095189,  5.19615242,  5.65685425,
        5.83095189,  5.74456265,  5.74456265,  9.05538514,  9.53939201,
        5.74456265,  6.92820323,  6.32455532,  9.94987437,  8.88819442,
        7.07106781,  7.        , 10.34408043,  8.18535277,  7.87400787,
        7.81024968,  9.64365076,  8.        ,  9.11043358, 11.        ,
       10.04987562,  8.88819442,  9.38083152,  8.54400375,  9.43398113,
       11.53256259, 11.61895004, 61.61980201, 31.27299154, 12.68857754,
        9.89949494, 61.68468205, 11.61895004, 26.70205985, 20.34698995,
       10.81665383, 61.76568627,  9.8488578 , 17.1464282 , 10.77032961,
       61.83041323, 12.36931688, 13.67479433, 16.64331698, 10.67707825,
       16.        , 11.78982612, 61.94352266, 14.10673598, 17.77638883,
       61.99193496, 40.55859958, 13.78404875, 12.36931688, 62.05642594,
       25.70992026, 11.61895004, 11.83215957, 12.76714533, 65.08

In [124]:
t_s = t_s ** 2

In [125]:
threshold = 17

count = np.sum(t_s < threshold)

count

np.int64(352)

In [126]:
top_secret = t_s.reshape(-1, 5, order='F')

top_secret

array([[5.900e+01, 1.440e+02, 3.610e+02, 9.000e+00, 2.050e+01],
       [5.000e+01, 1.000e+00, 1.521e+03, 7.500e+00, 3.050e+01],
       [2.900e+01, 5.290e+02, 2.250e+02, 1.050e+01, 3.000e+00],
       [9.000e+00, 3.610e+02, 2.250e+02, 3.500e+00, 9.000e+00],
       [2.900e+01, 3.721e+03, 1.600e+01, 4.000e+00, 5.000e-01],
       [7.300e+01, 1.521e+03, 3.721e+03, 3.050e+01, 1.000e+00],
       [1.600e+01, 6.400e+01, 1.000e+00, 5.000e-01, 5.000e+00],
       [3.400e+01, 1.000e+00, 5.290e+02, 7.000e+00, 7.500e+00],
       [2.700e+01, 1.521e+03, 6.400e+01, 2.000e+00, 1.050e+01],
       [3.200e+01, 3.721e+03, 8.100e+01, 3.050e+01, 9.500e+00],
       [3.400e+01, 9.000e+00, 1.440e+02, 1.950e+01, 3.050e+01],
       [3.300e+01, 1.000e+00, 2.500e+01, 4.000e+00, 2.000e+00],
       [3.300e+01, 1.521e+03, 3.721e+03, 9.000e+00, 5.000e-01],
       [8.200e+01, 9.000e+00, 8.100e+01, 7.500e+00, 1.250e+01],
       [9.100e+01, 6.400e+01, 1.960e+02, 1.050e+01, 2.750e+01],
       [3.300e+01, 3.025e+03, 3.721e+03,

In [116]:
# Extract even numbers from the top_secret array after rounding and converting to integer
top_secret_round = np.arange(2, 383, 2)

top_secret_round

array([  2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24,  26,
        28,  30,  32,  34,  36,  38,  40,  42,  44,  46,  48,  50,  52,
        54,  56,  58,  60,  62,  64,  66,  68,  70,  72,  74,  76,  78,
        80,  82,  84,  86,  88,  90,  92,  94,  96,  98, 100, 102, 104,
       106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130,
       132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156,
       158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182,
       184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208,
       210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234,
       236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260,
       262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286,
       288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312,
       314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338,
       340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 36

In [128]:
top_secret[: , 0] -= top_secret_round

top_secret

array([[5.700e+01, 1.440e+02, 3.610e+02, 9.000e+00, 2.050e+01],
       [4.600e+01, 1.000e+00, 1.521e+03, 7.500e+00, 3.050e+01],
       [2.300e+01, 5.290e+02, 2.250e+02, 1.050e+01, 3.000e+00],
       [1.000e+00, 3.610e+02, 2.250e+02, 3.500e+00, 9.000e+00],
       [1.900e+01, 3.721e+03, 1.600e+01, 4.000e+00, 5.000e-01],
       [6.100e+01, 1.521e+03, 3.721e+03, 3.050e+01, 1.000e+00],
       [2.000e+00, 6.400e+01, 1.000e+00, 5.000e-01, 5.000e+00],
       [1.800e+01, 1.000e+00, 5.290e+02, 7.000e+00, 7.500e+00],
       [9.000e+00, 1.521e+03, 6.400e+01, 2.000e+00, 1.050e+01],
       [1.200e+01, 3.721e+03, 8.100e+01, 3.050e+01, 9.500e+00],
       [1.200e+01, 9.000e+00, 1.440e+02, 1.950e+01, 3.050e+01],
       [9.000e+00, 1.000e+00, 2.500e+01, 4.000e+00, 2.000e+00],
       [7.000e+00, 1.521e+03, 3.721e+03, 9.000e+00, 5.000e-01],
       [5.400e+01, 9.000e+00, 8.100e+01, 7.500e+00, 1.250e+01],
       [6.100e+01, 6.400e+01, 1.960e+02, 1.050e+01, 2.750e+01],
       [1.000e+00, 3.025e+03, 3.721e+03,

In [129]:
top_secret[18:25 , 2] -= 100

top_secret

array([[5.700e+01, 1.440e+02, 3.610e+02, 9.000e+00, 2.050e+01],
       [4.600e+01, 1.000e+00, 1.521e+03, 7.500e+00, 3.050e+01],
       [2.300e+01, 5.290e+02, 2.250e+02, 1.050e+01, 3.000e+00],
       [1.000e+00, 3.610e+02, 2.250e+02, 3.500e+00, 9.000e+00],
       [1.900e+01, 3.721e+03, 1.600e+01, 4.000e+00, 5.000e-01],
       [6.100e+01, 1.521e+03, 3.721e+03, 3.050e+01, 1.000e+00],
       [2.000e+00, 6.400e+01, 1.000e+00, 5.000e-01, 5.000e+00],
       [1.800e+01, 1.000e+00, 5.290e+02, 7.000e+00, 7.500e+00],
       [9.000e+00, 1.521e+03, 6.400e+01, 2.000e+00, 1.050e+01],
       [1.200e+01, 3.721e+03, 8.100e+01, 3.050e+01, 9.500e+00],
       [1.200e+01, 9.000e+00, 1.440e+02, 1.950e+01, 3.050e+01],
       [9.000e+00, 1.000e+00, 2.500e+01, 4.000e+00, 2.000e+00],
       [7.000e+00, 1.521e+03, 3.721e+03, 9.000e+00, 5.000e-01],
       [5.400e+01, 9.000e+00, 8.100e+01, 7.500e+00, 1.250e+01],
       [6.100e+01, 6.400e+01, 1.960e+02, 1.050e+01, 2.750e+01],
       [1.000e+00, 3.025e+03, 3.721e+03,

In [130]:
top_secret[: , 3:5] *= 2

top_secret

array([[5.700e+01, 1.440e+02, 3.610e+02, 1.800e+01, 4.100e+01],
       [4.600e+01, 1.000e+00, 1.521e+03, 1.500e+01, 6.100e+01],
       [2.300e+01, 5.290e+02, 2.250e+02, 2.100e+01, 6.000e+00],
       [1.000e+00, 3.610e+02, 2.250e+02, 7.000e+00, 1.800e+01],
       [1.900e+01, 3.721e+03, 1.600e+01, 8.000e+00, 1.000e+00],
       [6.100e+01, 1.521e+03, 3.721e+03, 6.100e+01, 2.000e+00],
       [2.000e+00, 6.400e+01, 1.000e+00, 1.000e+00, 1.000e+01],
       [1.800e+01, 1.000e+00, 5.290e+02, 1.400e+01, 1.500e+01],
       [9.000e+00, 1.521e+03, 6.400e+01, 4.000e+00, 2.100e+01],
       [1.200e+01, 3.721e+03, 8.100e+01, 6.100e+01, 1.900e+01],
       [1.200e+01, 9.000e+00, 1.440e+02, 3.900e+01, 6.100e+01],
       [9.000e+00, 1.000e+00, 2.500e+01, 8.000e+00, 4.000e+00],
       [7.000e+00, 1.521e+03, 3.721e+03, 1.800e+01, 1.000e+00],
       [5.400e+01, 9.000e+00, 8.100e+01, 1.500e+01, 2.500e+01],
       [6.100e+01, 6.400e+01, 1.960e+02, 2.100e+01, 5.500e+01],
       [1.000e+00, 3.025e+03, 3.721e+03,

In [131]:
top_secret = top_secret.flatten(order='F')

top_secret

array([5.700e+01, 4.600e+01, 2.300e+01, 1.000e+00, 1.900e+01, 6.100e+01,
       2.000e+00, 1.800e+01, 9.000e+00, 1.200e+01, 1.200e+01, 9.000e+00,
       7.000e+00, 5.400e+01, 6.100e+01, 1.000e+00, 1.400e+01, 4.000e+00,
       6.100e+01, 3.900e+01, 8.000e+00, 5.000e+00, 6.100e+01, 1.900e+01,
       1.200e+01, 9.000e+00, 3.900e+01, 8.000e+00, 2.500e+01, 6.100e+01,
       3.900e+01, 1.500e+01, 2.200e+01, 5.000e+00, 1.900e+01, 6.100e+01,
       6.100e+01, 3.721e+03, 9.000e+02, 8.100e+01, 1.600e+01, 3.721e+03,
       4.900e+01, 6.250e+02, 3.240e+02, 2.500e+01, 3.721e+03, 1.000e+00,
       1.960e+02, 1.600e+01, 3.721e+03, 4.900e+01, 8.100e+01, 1.690e+02,
       4.000e+00, 1.440e+02, 2.500e+01, 3.721e+03, 8.100e+01, 1.960e+02,
       3.721e+03, 1.521e+03, 6.400e+01, 2.500e+01, 3.721e+03, 5.290e+02,
       1.000e+00, 4.000e+00, 2.500e+01, 4.096e+03, 3.721e+03, 7.290e+02,
       1.440e+02, 1.440e+02, 3.721e+03, 1.690e+02, 8.100e+01, 1.690e+02,
       3.610e+02, 6.250e+02, 3.721e+03, 5.290e+02, 

In [133]:
top_secret_100 = top_secret[500:]
print(np.max(top_secret_100))

63.99999999999695


In [134]:
top_secret[37:465] = np.sqrt(top_secret[37:465])

top_secret



array([57.        , 46.        , 23.        ,  1.        , 19.        ,
       61.        ,  2.        , 18.        ,  9.        , 12.        ,
       12.        ,  9.        ,  7.        , 54.        , 61.        ,
        1.        , 14.        ,  4.        , 61.        , 39.        ,
        8.        ,  5.        , 61.        , 19.        , 12.        ,
        9.        , 39.        ,  8.        , 25.        , 61.        ,
       39.        , 15.        , 22.        ,  5.        , 19.        ,
       61.        , 61.        , 61.        , 30.        ,  9.        ,
        4.        , 61.        ,  7.        , 25.        , 18.        ,
        5.        , 61.        ,  1.        , 14.        ,  4.        ,
       61.        ,  7.        ,  9.        , 13.        ,  2.        ,
       12.        ,  5.        , 61.        ,  9.        , 14.        ,
       61.        , 39.        ,  8.        ,  5.        , 61.        ,
       23.        ,  1.        ,  2.        ,  5.        , 64.  

In [135]:
top_secret = np.round(top_secret)

top_secret

array([57., 46., 23.,  1., 19., 61.,  2., 18.,  9., 12., 12.,  9.,  7.,
       54., 61.,  1., 14.,  4., 61., 39.,  8.,  5., 61., 19., 12.,  9.,
       39.,  8., 25., 61., 39., 15., 22.,  5., 19., 61., 61., 61., 30.,
        9.,  4., 61.,  7., 25., 18.,  5., 61.,  1., 14.,  4., 61.,  7.,
        9., 13.,  2., 12.,  5., 61.,  9., 14., 61., 39.,  8.,  5., 61.,
       23.,  1.,  2.,  5., 64., 61., 27., 12., 12., 61., 13.,  9., 13.,
       19., 25., 61., 23.,  5., 18.,  5., 61., 39.,  8.,  5., 61.,  2.,
       15., 18., 15.,  7., 15., 22.,  5., 19., 54., 61., 61., 61., 27.,
       14.,  4., 61., 39.,  8.,  5., 61., 13., 15., 13.,  5., 61., 18.,
        1., 39.,  8., 19., 61., 15., 21., 39.,  7., 18.,  1.,  2.,  5.,
       53., 61., 58., 28.,  5., 23.,  1., 18.,  5., 61., 39.,  8.,  5.,
       61., 36.,  1.,  2.,  2.,  5., 18., 23., 15.,  3., 11., 54., 61.,
       13., 25., 61., 19., 15., 14., 55., 61., 61., 61., 46.,  8.,  5.,
       61., 10.,  1., 23., 19., 61., 39.,  8.,  1., 39., 61.,  2

In [136]:
top_secret[top_secret == 39] = 20
top_secret

array([57., 46., 23.,  1., 19., 61.,  2., 18.,  9., 12., 12.,  9.,  7.,
       54., 61.,  1., 14.,  4., 61., 20.,  8.,  5., 61., 19., 12.,  9.,
       20.,  8., 25., 61., 20., 15., 22.,  5., 19., 61., 61., 61., 30.,
        9.,  4., 61.,  7., 25., 18.,  5., 61.,  1., 14.,  4., 61.,  7.,
        9., 13.,  2., 12.,  5., 61.,  9., 14., 61., 20.,  8.,  5., 61.,
       23.,  1.,  2.,  5., 64., 61., 27., 12., 12., 61., 13.,  9., 13.,
       19., 25., 61., 23.,  5., 18.,  5., 61., 20.,  8.,  5., 61.,  2.,
       15., 18., 15.,  7., 15., 22.,  5., 19., 54., 61., 61., 61., 27.,
       14.,  4., 61., 20.,  8.,  5., 61., 13., 15., 13.,  5., 61., 18.,
        1., 20.,  8., 19., 61., 15., 21., 20.,  7., 18.,  1.,  2.,  5.,
       53., 61., 58., 28.,  5., 23.,  1., 18.,  5., 61., 20.,  8.,  5.,
       61., 36.,  1.,  2.,  2.,  5., 18., 23., 15.,  3., 11., 54., 61.,
       13., 25., 61., 19., 15., 14., 55., 61., 61., 61., 46.,  8.,  5.,
       61., 10.,  1., 23., 19., 61., 20.,  8.,  1., 20., 61.,  2

In [140]:
top_secret_evens = top_secret[top_secret % 2 == 0]
print(len(top_secret_evens))

344


### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns. This conversion should be **column-wise**; that is, the first several elements of the array should all end up in the first **column** of your matrix.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array. Again this should be **column-wise**, so that the first column of the matrix becomes the first elements of the array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [138]:
''.join(map(str, symbols_df.loc[top_secret-1]['Symbol']))

'\'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe. "Beware the Jabberwock, my son!   The jaws that bite, the claws that catch! Beware the Jubjub bird, and shun   The frumious Bandersnatch!" He took his vorpal sword in hand;   Long time the manxome foe he sought- So rested he by the Tumtum tree   And stood awhile in tmought.)And, as in uffish thought he stood,   The Jabberwock, with eyes of flame, Came whiffling through the tulgey wood,   And burbled as it came! One, two! One, two! And through and through   The vorpal blade went snicker-snack! He left it dead, and with its head   He went galumphing back. "And hast thou slain the Jabberwock?   Come to my arms, my beamish boy! O frabjous day! Callooh! Callay!"   He chortled in his joy. \'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe.'